### DISCUSSION TITLE
Changing global variable?

### DISCUSSION THREAD

**Discussion by TomRui** [jspsych/jsPsych#1511] [Q&A] [answered]
Hi all,
I'm programming a task where the response window of the participant changes depending on the performance in the last block. For this, I'm using a global variable called adaptive_response_window which I want to change after every block of 18 trials. However, I find that I can't change this global variable at all and after doing all kinds of googling I still can't find out why.
Please find below a short piece of code to exemplify the problem. I expect the code to display 200 at the debugging trial because I tried to change the initial value using 1) a function call, and 2) on the finish of a trial. Instead it still displays the original value of 1230. Does anyone know what is going wrong here and how I can change the global variable?
```js
// initialize response window
var adaptive_response_window = 1230
// try to change variable using function call
function myfunc() {
adaptive_response_window = 200
}
var update_adaptive_window = {
type: 'call-function',
func: myfunc
}
// try to change variable on trial finish
var trial = {
type: 'html-keyboard-response',
//response_ends_trial: true,
stimulus: 'hi',
on_finish: function(){
adaptive_response_window = 200}
}
// display response window to see if it changed
var debugging = {
type: 'instructions',
pages: [
adaptive_response_window
],
key_forward: ' '
}
jsPsych.init({
timeline: [update_adaptive_window, trial, debugging],
});
```
Thanks a lot for any help anyone could provide! And please let me know if this isn't the right place to ask.

**Comment by becky-gilbert** [comment, marked as answer]
Hi @TomRui! Your functions that update the variable's value look like they should work. The problem might just be that you're not seeing the updated value because the debugging trial doesn't use a dynamic parameter for the page content, so it'll only ever show the initial value.
It might be easier to set up your debugging trial with the html-keyboard-response plugin. Another thing you can do is use `console.log` in your functions to see when they're called and check variable values. Try this:
```js
​// initialize response window​
​var​ ​adaptive_response_window​ ​=​ ​1230​
​// try to change variable using function call​
​function​ ​myfunc​(​)​ ​{​
  ​adaptive_response_window​ ​+= 200​;
  console.log('myfunc value: ', adaptive_response_window)
​}​
​var​ ​update_adaptive_window​ ​=​ ​{​
  ​type​: ​'call-function'​,​
  ​func​: ​myfunc​
​}​
​// try to change variable on trial finish​
​var​ ​trial​ ​=​ ​{​
  ​type​: ​'html-keyboard-response'​,​
  ​//response_ends_trial: true,​
  ​stimulus​: ​'hi'​,​
  ​on_finish​: ​function​(​)​{​
    ​adaptive_response_window​ ​+= ​200​;
    console.log('on finish value: ',adaptive_response_window);
  }​
​}​
// display response window to see if it changed​
​var​ ​debugging​ ​=​ ​{​
  ​type​: ​'html-keyboard-response'​,​
  stimulus: function () {
    return adaptive_response_window;
  }
​}​
​jsPsych​.​init​(​{​
  ​timeline​: ​[​update_adaptive_window​,​ ​trial​,​ ​debugging​]​,​
​}​)​;
```

**Reply by TomRui** [reply]
That's the problem indeed! When I replace all _adaptive_response_window_ variables with

function () {
    return adaptive_response_window;
 }

The tasks works as expected.

So if I understand correctly this forces it to use the dynamic parameter?

Thanks a lot for the help :)

**Reply by becky-gilbert** [reply]
Great! 
Yes, in most cases if you use a function as a parameter value then the function will run right before the trial starts, which allows you to make the parameter value dynamic. You can read about this in the [docs here](https://www.jspsych.org/overview/trial/#dynamic-parameters).

**Reply by TomRui** [reply]
Perfect, thanks a lot - and cheers for the link to the documentation, I must have missed that
